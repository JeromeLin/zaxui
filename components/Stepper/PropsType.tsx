interface BaseProps {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'circle';
  value?: number;
  defaultValue?: number;
  step: number;
  min: number;
  max: number;
  block?: boolean;
  disabled?: boolean;
  onInputChange?: (value: number) => void;
  onChange?: (value: number) => void;
}

export interface StepperProps extends BaseProps {
  prefixCls?: string;
  className?: string;
}
