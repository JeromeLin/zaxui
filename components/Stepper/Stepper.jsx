import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

function getValue(props, defaultValue) {
  if ('value' in props) {
    return props.value;
  }
  if ('defaultValue' in props) {
    return props.defaultValue;
  }
  return defaultValue;
}

class Stepper extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: getValue(props, 0),
      lastValue: getValue(props, 0),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: Number(getValue(nextProps, 0)),
        lastValue: Number(getValue(nextProps, 0)),
      });
    }
  }

  onInputChange(value) {
    const { onChange } = this.props;
    this.setState({
      value,
    });
    typeof onChange === 'function' && onChange(value);
  }

  onInputBlur(value) {
    const { min, max, onBlur } = this.props;
    if (value === '' || isNaN(value)) {
      value = this.state.lastValue;
    }
    if (min !== null && value < min) {
      value = min;
    }
    if (max !== null && value > max) {
      value = max;
    }
    this.setState({
      value,
      lastValue: value,
    });
    typeof onBlur === 'function' && onBlur(value);
  }

  onSubClick() {
    const { step } = this.props;
    const value = Number(this.state.value) - step;
    this.onInputChange(value);
  }

  onPlusClick() {
    const { step } = this.props;
    const value = Number(this.state.value) + step;
    this.onInputChange(value);
  }

  render() {
    const { prefixCls, className, theme, size, shape, disabled, min, max } = this.props;
    const { value } = this.state;

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
      [`shape-${shape}`]: !!shape,
      disabled,
    });

    const subDisabled = !!(typeof min === 'number' && value <= min) || disabled;
    const plusDisabled = !!(typeof max === 'number' && value >= max) || disabled;

    const subCls = classnames({
      [`${prefixCls}-sub`]: true,
      disabled: subDisabled,
    });

    const plusCls = classnames({
      [`${prefixCls}-plus`]: true,
      disabled: plusDisabled,
    });

    return (
      <span className={cls}>
        <span className={subCls} onClick={() => !subDisabled && this.onSubClick()}><Icon type="minus" /></span>
        <input className={`${prefixCls}-body`} type="tel" value={value} onChange={e => this.onInputChange(e.target.value)} onBlur={e => this.onInputBlur(e.target.value)} disabled={disabled} />
        <span className={plusCls} onClick={() => !plusDisabled && this.onPlusClick()}><Icon type="add" /></span>
      </span>
    );
  }
}

Stepper.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  size: PropTypes.oneOf(['xl', 'lg', 'sm', 'xs']),
  shape: PropTypes.oneOf(['radius', 'round', 'circle']),
  disabled: PropTypes.bool,
  step: PropTypes.number,
  min: PropTypes.number,
  max: PropTypes.number,
};

Stepper.defaultProps = {
  prefixCls: 'za-stepper',
  className: null,
  theme: 'primary',
  size: null,
  shape: null,
  disabled: false,
  step: 1,
  min: null,
  max: null,
};

export default Stepper;
