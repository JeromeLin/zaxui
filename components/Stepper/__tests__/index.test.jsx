import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Stepper from '../index';

describe('Stepper', () => {
  it('renders correctly', () => {
    const onChange = jest.fn();

    const wrapper = render(
      <Stepper onChange={onChange} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
