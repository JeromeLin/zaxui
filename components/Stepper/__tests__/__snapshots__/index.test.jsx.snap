// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper defaultValue 1`] = `
<span
  className="za-stepper theme-primary"
>
  <span
    className="za-stepper-sub"
    onClick={[Function]}
  >
    <Icon
      className={null}
      prefixCls="za-icon"
      theme={null}
      type="minus"
    />
  </span>
  <input
    className="za-stepper-body"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    type="tel"
    value={2}
  />
  <span
    className="za-stepper-plus"
    onClick={[Function]}
  >
    <Icon
      className={null}
      prefixCls="za-icon"
      theme={null}
      type="add"
    />
  </span>
</span>
`;

exports[`Stepper renders correctly 1`] = `
<span
  class="za-stepper theme-primary"
>
  <span
    class="za-stepper-sub"
  >
    <i
      class="za-icon za-icon-minus"
    />
  </span>
  <input
    class="za-stepper-body"
    type="tel"
    value="0"
  />
  <span
    class="za-stepper-plus"
  >
    <i
      class="za-icon za-icon-add"
    />
  </span>
</span>
`;
